src/execution/radix_partitioned_hashtable.cpp

```diff
@@ -302,22 +323,244 @@ class RadixAggregateFinalizeTask : public ExecutorTask {
+
+	gstate.partition_info = std::move(new_partition_info);
+	InitializeFinalizedHTs(executor.context, state);
+}
+
+bool RadixPartitionedHashTable::ForceSingleHT(GlobalSinkState &state) {
+	auto &gstate = state.Cast<RadixHTGlobalState>();
+	return gstate.partition_info->n_partitions < 2;
+}
+
+bool RadixPartitionedHashTable::AnyPartitioned(GlobalSinkState &state) {
+	auto &gstate = state.Cast<RadixHTGlobalState>();
+	for (auto &pht : gstate.intermediate_hts) {
+		if (pht->IsPartitioned()) {
+			return true;
+		}
 	}
+	return false;
 }
 
-bool RadixPartitionedHashTable::ForceSingleHT(GlobalSinkState &state) const {
+void RadixPartitionedHashTable::GetRepartitionInfo(ClientContext &context, GlobalSinkState &state,
+                                                   idx_t &repartition_radix_bits, idx_t &concurrent_repartitions,
+                                                   idx_t &tasks_per_partition) {
 	auto &gstate = state.Cast<RadixHTGlobalState>();
-	return gstate.partition_info.n_partitions < 2;
+	const auto num_partitions = gstate.partition_info->n_partitions;
+	const auto radix_bits = gstate.partition_info->radix_bits;
+	D_ASSERT(IsPowerOfTwo(num_partitions));
+
+	vector<idx_t> partition_counts(num_partitions, 0);
+	vector<idx_t> partition_sizes(num_partitions, 0);
+	for (const auto &ht : gstate.intermediate_hts) {
+		for (idx_t partition_idx = 0; partition_idx < num_partitions; partition_idx++) {
+			partition_counts[partition_idx] += ht->GetPartitionCount(partition_idx);
+			partition_sizes[partition_idx] += ht->GetPartitionSize(partition_idx);
+		}
+	}
+
+	idx_t total_size = 0;
+	idx_t max_partition_idx = 0;
+	idx_t max_partition_size = 0;
+	for (idx_t partition_idx = 0; partition_idx < num_partitions; partition_idx++) {
+		const auto &partition_count = partition_counts[partition_idx];
+		const auto &partition_size = partition_sizes[partition_idx];
+		auto partition_ht_size =
+		    partition_size + GroupedAggregateHashTable::FirstPartSize(partition_count, HtEntryType::HT_WIDTH_64);
+		if (partition_ht_size > max_partition_size) {
+			max_partition_idx = partition_idx;
+			max_partition_size = partition_ht_size;
+		}
+		total_size += partition_ht_size;
+	}
+
+	// Switch to out-of-core finalize at ~60%
+	const auto max_ht_size = double(0.6) * BufferManager::GetBufferManager(context).GetMaxMemory();
+	const idx_t n_threads = PreviousPowerOfTwo(TaskScheduler::GetScheduler(context).NumberOfThreads());
+	D_ASSERT(IsPowerOfTwo(n_threads));
+	if (!context.config.force_external && total_size < max_ht_size) {
+		// In-memory finalize
+		if (num_partitions >= n_threads) {
+			// Can already keep all threads busy
+			repartition_radix_bits = radix_bits;
+			tasks_per_partition = 1;
+		} else {
+			// Repartition to keep all threads busy
+			repartition_radix_bits = RadixPartitioning::RadixBits(NextPowerOfTwo(n_threads));
+			tasks_per_partition = n_threads / num_partitions;
```

Comment:
If n_threads == num_partitions + 1 then won't a thread be idle?

Reply:
No, then it will start repartitioning the next partition